### PR TITLE
Release/v1.1.0

### PR DESCRIPTION
**Pull request by santidrj**

### CODE REVIEW

File: src/main/java/org/pesmypetcare/webservice/dao/PetDaoImpl.java [hunk 1]
```diff
@@ -26,27 +26,47 @@
...
+        return storageDao;
+    }
+
     @Override
     public void createPet(String owner, String name, PetEntity petEntity) {
-        DocumentReference petRef = usersRef.document(owner).collection(PETS_KEY).document(name);
+        DocumentReference petRef = usersRef.document(owner).
+                collection(PETS_KEY).document(name);
         petRef.set(petEntity);
+        System.out.println("Pet created");
```
Comment (AlvarTB) [inline src/main/java/org/pesmypetcare/webservice/dao/PetDaoImpl.java]: I think you should delete this line. It's a println anyways.

File: src/main/java/org/pesmypetcare/webservice/service/MedicationServiceImp.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+import org.pesmypetcare.webservice.error.DatabaseAccessException;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+
+@Service
+public class MedicationServiceImp implements MedicationService {
```
Comment (oricat8) [inline src/main/java/org/pesmypetcare/webservice/service/MedicationServiceImp.java]: I think that this document has to have the "Impl" termination despite Imp.

File: src/test/java/org/pesmypetcare/webservice/dao/WeightDaoTest.java [hunk 3]
```diff
@@ -0,0 +1,386 @@
...
+        given(weightsRef.get()).willReturn(futureQuery);
+        given(futureQuery.get()).willReturn(querySnapshot);
+        given(querySnapshot.getDocuments()).willReturn(weightsDocuments);
+        given(weightsDocuments.iterator()).willReturn(it);
+        given(it.hasNext()).willReturn(true);
+        given(it.hasNext()).willReturn(true);
+        given(it.hasNext()).willReturn(false);
+
+        List<Map<String, Object>> list = weightDao.getAllWeight(owner, petName);
+        assertEquals(weightList, list, "Should return a List containing all meals Data");
```
Comment (AlvarTB) [inline src/test/java/org/pesmypetcare/webservice/dao/WeightDaoTest.java]: "All weights data"?

File: src/test/java/org/pesmypetcare/webservice/dao/WeightDaoTest.java [hunk 4]
```diff
@@ -0,0 +1,386 @@
...
+        given(futureQuery.get()).willReturn(querySnapshot);
+        given(querySnapshot.getDocuments()).willReturn(weightsDocuments);
+        given(weightsDocuments.iterator()).willReturn(it);
+        given(it.hasNext()).willReturn(true);
+        given(it.hasNext()).willReturn(true);
+        given(it.hasNext()).willReturn(false);
+
+        List<Map<String, Object>> list = weightDao.getAllWeightsBetween(owner, petName, date, date2);
+
+        assertEquals(weightList, list, "Should return a List containing all meals between two dates");
```
Comment (AlvarTB) [inline src/test/java/org/pesmypetcare/webservice/dao/WeightDaoTest.java]: "All weights between two dates"?

**Review by oricat8 — CHANGES_REQUESTED** [review]

**Review by AlvarTB — CHANGES_REQUESTED** [review]

File: src/main/java/org/pesmypetcare/webservice/service/MedicationServiceImp.java (hunk 2 above)
Comment (santidrj) [reply src/main/java/org/pesmypetcare/webservice/service/MedicationServiceImp.java]: Done

File: src/main/java/org/pesmypetcare/webservice/dao/PetDaoImpl.java (hunk 1 above)
Comment (santidrj) [reply src/main/java/org/pesmypetcare/webservice/dao/PetDaoImpl.java]: Done

File: src/test/java/org/pesmypetcare/webservice/dao/WeightDaoTest.java (hunk 3 above)
Comment (santidrj) [reply src/test/java/org/pesmypetcare/webservice/dao/WeightDaoTest.java]: Resolved

File: src/test/java/org/pesmypetcare/webservice/dao/WeightDaoTest.java (hunk 4 above)
Comment (santidrj) [reply src/test/java/org/pesmypetcare/webservice/dao/WeightDaoTest.java]: Resolved

File: src/main/java/org/pesmypetcare/webservice/service/MedicationServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    }
+
+    @Override
+    public Object getMedicationField(String owner, String petName, String date, String name, String field)
+            throws DatabaseAccessException {
+        return medicationDao.getMedicationField(owner, petName, date, name, field);
+    }
+
+    @Override
+    public void updateMedicationField(String owner, String petName, String date, String name, String field,
```
Comment (santidrj) [inline src/main/java/org/pesmypetcare/webservice/service/MedicationServiceImpl.java:60]: Codacy Issue found: [More than 5 parameters (found 6).](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5426774)

**Comment by santidrj** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- src/main/java/org/pesmypetcare/webservice/firebaseservice/CalendarServiceFactory.java  2
- src/test/java/org/pesmypetcare/webservice/service/WeekTrainingServiceTest.java  1
- src/main/java/org/pesmypetcare/webservice/controller/appmanager/StorageRestController.java  1
- src/test/java/org/pesmypetcare/webservice/dao/FreqWashDaoTest.java  1
- src/test/java/org/pesmypetcare/webservice/controller/petmanager/PetWeightRestControllerTest.java  1
- src/test/java/org/pesmypetcare/webservice/dao/KcalDaoTest.java  1
- src/main/java/org/pesmypetcare/webservice/controller/petmanager/FreqWashRestController.java  1
- src/test/java/org/pesmypetcare/webservice/dao/WeekTrainingDaoTest.java  1
- src/main/java/org/pesmypetcare/webservice/controller/usermanager/UserRestController.java  1
- src/test/java/org/pesmypetcare/webservice/dao/FreqTrainingDaoTest.java  1
- src/main/java/org/pesmypetcare/webservice/service/FreqWashServiceImpl.java  1
- src/main/java/org/pesmypetcare/webservice/controller/petmanager/KcalAverageRestController.java  1
- src/main/java/org/pesmypetcare/webservice/service/StorageServiceImpl.java  1
- src/main/java/org/pesmypetcare/webservice/controller/petmanager/PetMedicationRestController.java  1
- src/test/java/org/pesmypetcare/webservice/controller/petmanager/PetWeekTrainingRestControllerTest.java  1
- src/test/java/org/pesmypetcare/webservice/service/FreqWashServiceTest.java  1
- src/main/java/org/pesmypetcare/webservice/service/WeekTrainingServiceImpl.java  1
- src/test/java/org/pesmypetcare/webservice/service/FreqTrainingServiceTest.java  1
- src/main/java/org/pesmypetcare/webservice/service/MedicationServiceImpl.java  1
- src/main/java/org/pesmypetcare/webservice/dao/StorageDaoImpl.java  2
- src/test/java/org/pesmypetcare/webservice/controller/petmanager/PetKcalAverageRestControllerTest.java  1
- src/main/java/org/pesmypetcare/webservice/error/InitializationException.java  1
- src/main/java/org/pesmypetcare/webservice/service/FreqTrainingServiceImpl.java  1
- src/main/java/org/pesmypetcare/webservice/form/StorageForm.java  1
- src/main/java/org/pesmypetcare/webservice/service/MealServiceImpl.java  1
- src/test/java/org/pesmypetcare/webservice/dao/PetDaoTest.java  2
- src/test/java/org/pesmypetcare/webservice/service/KcalServiceTest.java  1
- src/test/java/org/pesmypetcare/webservice/dao/KcalAverageDaoTest.java  1
- src/main/java/org/pesmypetcare/webservice/controller/petmanager/WeekTrainingRestController.java  1
- src/test/java/org/pesmypetcare/webservice/service/KcalAverageServiceTest.java  1
- src/test/java/org/pesmypetcare/webservice/controller/appmanager/StorageRestControllerTest.java  1
- src/main/java/org/pesmypetcare/webservice/controller/petmanager/PetMealRestController.java  1
- src/main/java/org/pesmypetcare/webservice/service/KcalAverageServiceImpl.java  1
- src/main/java/org/pesmypetcare/webservice/service/WeightServiceImpl.java  1
- src/test/java/org/pesmypetcare/webservice/service/StorageServiceTest.java  1
- src/main/java/org/pesmypetcare/webservice/service/GoogleCalendarServiceImpl.java  2
- src/main/java/org/pesmypetcare/webservice/jsonhandler/DateTimeHandler.java  2
- src/main/java/org/pesmypetcare/webservice/error/CalendarAccessException.java  1
- src/main/java/org/pesmypetcare/webservice/controller/petmanager/WeightRestController.java  1
- src/main/java/org/pesmypetcare/webservice/controller/petmanager/FreqTrainingRestController.java  1
- src/test/java/org/pesmypetcare/webservice/service/MedicationServiceTest.java  1
- src/test/java/org/pesmypetcare/webservice/controller/petmanager/PetFreqWashRestControllerTest.java  1
- src/main/java/org/pesmypetcare/webservice/controller/petmanager/KcalRestController.java  1
- src/test/java/org/pesmypetcare/webservice/controller/petmanager/PetFreqTrainingRestControllerTest.java  1
- src/test/java/org/pesmypetcare/webservice/dao/MedicationDaoTest.java  1
- src/test/java/org/pesmypetcare/webservice/dao/StorageDaoTest.java  1
- src/main/java/org/pesmypetcare/webservice/dao/WeekTrainingDaoImpl.java  5
- src/main/java/org/pesmypetcare/webservice/service/KcalServiceImpl.java  1
- src/test/java/org/pesmypetcare/webservice/controller/petmanager/PetKcalRestControllerTest.java  1
         

Clones added
============
- src/main/java/org/pesmypetcare/webservice/dao/MealDao.java  1
- src/test/java/org/pesmypetcare/webservice/dao/FreqWashDaoTest.java  42
- src/test/java/org/pesmypetcare/webservice/dao/KcalDaoTest.java  42
- src/main/java/org/pesmypetcare/webservice/service/FreqTrainingService.java  1
- src/main/java/org/pesmypetcare/webservice/service/WeekTrainingService.java  1
- src/test/java/org/pesmypetcare/webservice/dao/WeekTrainingDaoTest.java  42
- src/main/java/org/pesmypetcare/webservice/service/FreqWashService.java  1
- src/test/java/org/pesmypetcare/webservice/dao/FreqTrainingDaoTest.java  42
- src/main/java/org/pesmypetcare/webservice/dao/MedicationDao.java  1
- src/test/java/org/pesmypetcare/webservice/dao/UserDaoTest.java  4
- src/main/java/org/pesmypetcare/webservice/service/MedicationService.java  1
- src/main/java/org/pesmypetcare/webservice/dao/WeightDao.java  1
- src/main/java/org/pesmypetcare/webservice/service/KcalAverageService.java  1
- src/main/java/org/pesmypetcare/webservice/dao/WeekTrainingDao.java  1
- src/test/java/org/pesmypetcare/webservice/dao/KcalAverageDaoTest.java  42
- src/main/java/org/pesmypetcare/webservice/dao/KcalDao.java  1
- src/main/java/org/pesmypetcare/webservice/service/KcalService.java  1
- src/main/java/org/pesmypetcare/webservice/dao/KcalAverageDao.java  1
- src/main/java/org/pesmypetcare/webservice/service/MealService.java  1
- src/main/java/org/pesmypetcare/webservice/dao/FreqWashDao.java  1
- src/main/java/org/pesmypetcare/webservice/service/WeightService.java  1
- src/main/java/org/pesmypetcare/webservice/dao/FreqTrainingDao.java  1
- src/test/java/org/pesmypetcare/webservice/dao/MedicationDaoTest.java  61
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5426774&bid=17796334)

**Review by marclll — APPROVED** [review]

**Review by AlvarTB — APPROVED** [review]

**Review by oricat8 — APPROVED** [review]

**Comment by santidrj** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Coverage increased per file
===========================
+ src/main/java/org/pesmypetcare/webservice/error/RestExceptionHandler.java  13
+ src/main/java/org/pesmypetcare/webservice/error/DatabaseAccessException.java  25
+ src/main/java/org/pesmypetcare/webservice/error/ErrorBody.java  50
+ src/main/java/org/pesmypetcare/webservice/firebaseservice/FirebaseFactory.java  1
         

Coverage decreased per file
===========================
- src/main/java/org/pesmypetcare/webservice/controller/usermanager/UserRestController.java  -12
- src/main/java/org/pesmypetcare/webservice/service/UserServiceImpl.java  -24
         

Complexity increasing per file
==============================
- src/main/java/org/pesmypetcare/webservice/controller/petmanager/FreqTrainingRestController.java  1
- src/main/java/org/pesmypetcare/webservice/controller/usermanager/UserRestController.java  1
- src/test/java/org/pesmypetcare/webservice/service/FreqWashServiceTest.java  1
- src/test/java/org/pesmypetcare/webservice/dao/StorageDaoTest.java  1
- src/main/java/org/pesmypetcare/webservice/dao/WeekTrainingDaoImpl.java  5
- src/main/java/org/pesmypetcare/webservice/controller/petmanager/FreqWashRestController.java  1
- src/test/java/org/pesmypetcare/webservice/controller/appmanager/StorageRestControllerTest.java  1
- src/main/java/org/pesmypetcare/webservice/firebaseservice/CalendarServiceFactory.java  2
- src/main/java/org/pesmypetcare/webservice/form/StorageForm.java  1
- src/test/java/org/pesmypetcare/webservice/dao/KcalDaoTest.java  1
- src/test/java/org/pesmypetcare/webservice/service/FreqTrainingServiceTest.java  1
- src/test/java/org/pesmypetcare/webservice/dao/FreqTrainingDaoTest.java  1
- src/test/java/org/pesmypetcare/webservice/dao/KcalAverageDaoTest.java  1
- src/main/java/org/pesmypetcare/webservice/service/WeightServiceImpl.java  1
- src/main/java/org/pesmypetcare/webservice/controller/appmanager/StorageRestController.java  1
- src/main/java/org/pesmypetcare/webservice/controller/petmanager/KcalRestController.java  1
- src/test/java/org/pesmypetcare/webservice/controller/petmanager/PetFreqTrainingRestControllerTest.java  1
- src/main/java/org/pesmypetcare/webservice/service/MealServiceImpl.java  1
- src/main/java/org/pesmypetcare/webservice/service/KcalServiceImpl.java  1
- src/test/java/org/pesmypetcare/webservice/controller/petmanager/PetWeekTrainingRestControllerTest.java  1
- src/test/java/org/pesmypetcare/webservice/dao/FreqWashDaoTest.java  1
- src/main/java/org/pesmypetcare/webservice/error/CalendarAccessException.java  1
- src/test/java/org/pesmypetcare/webservice/controller/petmanager/PetKcalAverageRestControllerTest.java  1
- src/main/java/org/pesmypetcare/webservice/service/StorageServiceImpl.java  1
- src/main/java/org/pesmypetcare/webservice/controller/petmanager/PetMealRestController.java  1
- src/test/java/org/pesmypetcare/webservice/service/KcalAverageServiceTest.java  1
- src/main/java/org/pesmypetcare/webservice/service/KcalAverageServiceImpl.java  1
- src/test/java/org/pesmypetcare/webservice/dao/WeekTrainingDaoTest.java  1
- src/test/java/org/pesmypetcare/webservice/service/WeekTrainingServiceTest.java  1
- src/main/java/org/pesmypetcare/webservice/error/InitializationException.java  1
- src/main/java/org/pesmypetcare/webservice/dao/StorageDaoImpl.java  2
- src/test/java/org/pesmypetcare/webservice/controller/petmanager/PetFreqWashRestControllerTest.java  1
- src/test/java/org/pesmypetcare/webservice/service/StorageServiceTest.java  1
- src/main/java/org/pesmypetcare/webservice/jsonhandler/DateTimeHandler.java  2
- src/test/java/org/pesmypetcare/webservice/service/KcalServiceTest.java  1
- src/main/java/org/pesmypetcare/webservice/controller/petmanager/WeekTrainingRestController.java  1
- src/main/java/org/pesmypetcare/webservice/controller/petmanager/WeightRestController.java  1
- src/main/java/org/pesmypetcare/webservice/service/GoogleCalendarServiceImpl.java  2
- src/main/java/org/pesmypetcare/webservice/controller/petmanager/KcalAverageRestController.java  1
- src/test/java/org/pesmypetcare/webservice/controller/petmanager/PetWeightRestControllerTest.java  1
- src/test/java/org/pesmypetcare/webservice/dao/PetDaoTest.java  2
- src/test/java/org/pesmypetcare/webservice/service/MedicationServiceTest.java  1
- src/test/java/org/pesmypetcare/webservice/dao/MedicationDaoTest.java  1
- src/test/java/org/pesmypetcare/webservice/controller/petmanager/PetKcalRestControllerTest.java  1
- src/main/java/org/pesmypetcare/webservice/service/FreqWashServiceImpl.java  1
- src/main/java/org/pesmypetcare/webservice/controller/petmanager/PetMedicationRestController.java  1
- src/main/java/org/pesmypetcare/webservice/service/MedicationServiceImpl.java  1
- src/main/java/org/pesmypetcare/webservice/service/WeekTrainingServiceImpl.java  1
- src/main/java/org/pesmypetcare/webservice/service/FreqTrainingServiceImpl.java  1
         

Clones added
============
- src/main/java/org/pesmypetcare/webservice/dao/KcalDao.java  1
- src/test/java/org/pesmypetcare/webservice/dao/KcalDaoTest.java  42
- src/test/java/org/pesmypetcare/webservice/dao/FreqTrainingDaoTest.java  42
- src/test/java/org/pesmypetcare/webservice/dao/KcalAverageDaoTest.java  42
- src/main/java/org/pesmypetcare/webservice/service/FreqWashService.java  1
- src/main/java/org/pesmypetcare/webservice/service/KcalService.java  1
- src/test/java/org/pesmypetcare/webservice/dao/FreqWashDaoTest.java  42
- src/main/java/org/pesmypetcare/webservice/service/FreqTrainingService.java  1
- src/main/java/org/pesmypetcare/webservice/dao/MealDao.java  1
- src/main/java/org/pesmypetcare/webservice/dao/WeekTrainingDao.java  1
- src/test/java/org/pesmypetcare/webservice/dao/WeekTrainingDaoTest.java  42
- src/main/java/org/pesmypetcare/webservice/dao/MedicationDao.java  1
- src/main/java/org/pesmypetcare/webservice/service/WeekTrainingService.java  1
- src/main/java/org/pesmypetcare/webservice/dao/FreqWashDao.java  1
- src/main/java/org/pesmypetcare/webservice/service/MealService.java  1
- src/main/java/org/pesmypetcare/webservice/dao/FreqTrainingDao.java  1
- src/main/java/org/pesmypetcare/webservice/service/MedicationService.java  1
- src/main/java/org/pesmypetcare/webservice/dao/KcalAverageDao.java  1
- src/main/java/org/pesmypetcare/webservice/service/WeightService.java  1
- src/test/java/org/pesmypetcare/webservice/dao/UserDaoTest.java  4
- src/main/java/org/pesmypetcare/webservice/dao/WeightDao.java  1
- src/test/java/org/pesmypetcare/webservice/dao/MedicationDaoTest.java  61
- src/main/java/org/pesmypetcare/webservice/service/KcalAverageService.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5426774&bid=17796334)

**Review by marclll — APPROVED** [review]

**Pull request merged**